pkg/agent/puller_test.go

```diff
@@ -0,0 +1,94 @@
+package agent
+
+import (
+	"sync"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"go.uber.org/zap"
+)
+
+func TestEmptyCommands(t *testing.T) {
+	logger := zap.NewNop().Sugar()
+	downloader := &Downloader{}
+	puller := Puller{
+		channelMap:  make(map[string]*ModelChannel),
+		completions: make(chan *ModelOp, 4),
+		opStats:     make(map[string]map[OpType]int),
+		waitGroup:   WaitGroupWrapper{sync.WaitGroup{}},
+		Downloader:  downloader,
+		logger:      logger,
+	}
```

Comment:
These seem a bit repetitive. Perhaps refactoring a bit and run different test cases in a loop?

Reply:
hey @terrytangyuan that's actually a great idea. I refactored them to be table-driven test, which cover the same assertions and validations for each of the separate test functions.